src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs

```diff
@@ -717,34 +718,89 @@ static void UpdateProjectReferences(
         }
 
         static ProjectUpdateState UpdateAnalyzerReferences(
-            ProjectId projectId,
+            Project projectBeforeMutation,
             SolutionChangeAccumulator solutionChanges,
             ProjectUpdateState projectUpdateState,
-            List<AnalyzerFileReference> analyzersRemovedInBatch,
-            List<AnalyzerFileReference> analyzersAddedInBatch)
+            List<string> analyzersRemovedInBatch,
+            List<string> analyzersAddedInBatch)
         {
-            // Analyzer reference removing...
-            if (analyzersRemovedInBatch.Count > 0)
-            {
-                projectUpdateState = projectUpdateState.WithIncrementalAnalyzerReferencesRemoved(analyzersRemovedInBatch);
+            if (analyzersRemovedInBatch.Count == 0 && analyzersAddedInBatch.Count == 0)
+                return projectUpdateState;
 
-                foreach (var analyzerReference in analyzersRemovedInBatch)
-                    solutionChanges.UpdateSolutionForProjectAction(projectId, solutionChanges.Solution.RemoveAnalyzerReference(projectId, analyzerReference));
-            }
+            var projectId = projectBeforeMutation.Id;
 
-            // Analyzer reference adding...
-            if (analyzersAddedInBatch.Count > 0)
-            {
-                projectUpdateState = projectUpdateState.WithIncrementalAnalyzerReferencesAdded(analyzersAddedInBatch);
+            // Use shared helper to figure out the new forked state.
+            var (newSolution, newProjectUpdateState) = UpdateProjectAnalyzerReferences(
+                solutionChanges.Solution, projectId, projectUpdateState, analyzersRemovedInBatch, analyzersAddedInBatch);
 
-                solutionChanges.UpdateSolutionForProjectAction(
-                    projectId, solutionChanges.Solution.AddAnalyzerReferences(projectId, analyzersAddedInBatch));
-            }
+            solutionChanges.UpdateSolutionForProjectAction(projectId, newSolution);
 
             return projectUpdateState;
         }
     }
 
+    public static (Solution newSolution, ProjectUpdateState newProjectUpdateState) UpdateProjectAnalyzerReferences(
+        Solution solution,
+        ProjectId projectId,
+        ProjectUpdateState projectUpdateState,
+        List<string> analyzersRemoved,
+        List<string> analyzersAdded)
+    {
+        Contract.ThrowIfTrue(analyzersRemoved.Count == 0 && analyzersAdded.Count == 0, "Should only be called when there is work to do");
+
+        // NOTE: Create the initial AnalyzerFileReferences for the analyzers we're adding with a shared shadow copy
+        // loader.  This is fine as we're just creating these to pass into CreateIsolatedAnalyzerReferencesAsync which
+        // will properly give them an isolated ALC to use instead.
+        var assemblyLoaderProvider = solution.Services.GetRequiredService<IAnalyzerAssemblyLoaderProvider>();
+        var sharedShadowCopyLoader = assemblyLoaderProvider.SharedShadowCopyLoader;
```

Comment:
in the VS process (netfx), if you try to load, you'll get the *original* assembly/types that were loaded the first time.  so you never get the new assemblies/types.

> And nothing in this path is still triggering a load?

not that i've been able to hit.  everything should be going to OOP.

--

Note: i have tested this with OOP off.  Reloading doesn't work.  But nothing breaks.   We make all these new references.  However, because they go through the same loader you get all the old assemblies/types/analyzers/source-generators.

Reply:
Is there a follow-up change that would actually block loading in-proc? It's a breaking change admittedly for third parties if anybody is using that, but it'd ensure we don't make mistakes going forward.